[Clang][AArch64] Change SME attributes for shared/new/preserved state.

sdesmalen-arm opened the pull request:
This patch replaces the `__arm_new_za`, `__arm_shared_za` and `__arm_preserves_za` attributes in favour of:
* `__arm_new("za")`
* `__arm_in("za")`
* `__arm_out("za")`
* `__arm_inout("za")`
* `__arm_preserves("za")`

As described in https://github.com/ARM-software/acle/pull/276.

One change is that `__arm_in/out/inout/preserves(S)` are all mutually exclusive, whereas previously it was fine to write `__arm_shared_za __arm_preserves_za`. This case is now represented with `__arm_in("za")`.

The current implementation uses the same LLVM attributes under the hood, since `__arm_in/out/inout` are all variations of "shared ZA", so can use the existing `aarch64_pstate_za_shared` attribute in LLVM.

A future patch will add support for the new "zt0" state as introduced with SME2.

sdesmalen-arm committed:
[Clang][AArch64] Change SME attributes for shared/new/preserved state.

This patch replaces the __arm_new_za, __arm_shared_za and __arm_preserves_za
attributes in favour of:
* `__arm_new("za")`
* `__arm_in("za")`
* `__arm_out("za")`
* `__arm_inout("za")`
* `__arm_preserves("za")`

As described in https://github.com/ARM-software/acle/pull/276.

One change is that `__arm_in/out/inout/preserves(S)` are all mutually exclusive,
whereas previously it was fine to write `__arm_shared_za __arm_preserves_za`.
This case is now represented with `__arm_in("za")`.

The current implementation uses the same LLVM attributes under the hood,
since `__arm_in/out/inout` are all variations of "shared ZA", so can
use the existing `aarch64_pstate_za_shared` attribute in LLVM.

A future patch will add support for the new "zt0" state as introduced with SME2.

sdesmalen-arm commented:
I'm aware the PR is quite large, but I didn't really know a sensible way to split this up. I'm happy to split things out if there are any suggestions.

rsandifo-arm commented on clang/include/clang/AST/Type.h:
```diff
@@ -4033,12 +4033,27 @@ class FunctionType : public Type {
-                                  // bitmask in FunctionTypeExtraBitfields.
+
+    // Describes the value of the state using ArmStateValue.
+    SME_PstateZAShift = 2,
```
It might make sense for these to be just `SME_ZA…`, rather than `SME_PstateZA…`, since they're now specifically about the contents of ZA (rather than PSTATE.ZA).

rsandifo-arm commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -6852,30 +6852,73 @@ without changing modes.
-The ``__arm_shared_za`` keyword applies to prototyped function types and specifies
-that the function shares SME's matrix storage (ZA) with its caller.  This
-means that:
+The ``__arm_in`` keyword applies to prototyped function types and specifies
```
This is referring to S before it has been defined.

rsandifo-arm commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -3696,6 +3696,12 @@ def err_sme_definition_using_sm_in_non_sme_target : Error<
   "function executed in streaming-SVE mode requires 'sme'">;
 def err_sme_definition_using_za_in_non_sme_target : Error<
   "function using ZA state requires 'sme'">;
+def err_conflicting_attributes_sme_state : Error<
```
`arm_state` might be better than `sme_state` (as for the errors below).  The new keywords aren't in principle specific to SME.

rsandifo-arm commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -6787,6 +6787,9 @@ void Parser::ParseDirectDeclarator(Declarator &D) {
       // For consistency with attribute parsing.
       Diag(Tok, diag::err_keyword_not_allowed) << Tok.getIdentifierInfo();
       ConsumeToken();
+      BalancedDelimiterTracker T(*this, tok::l_paren);
```
I think this needs to be conditional on whether the attribute accepts arguments.  In some contexts it's possible for an argumentless keyword like `__arm_streaming` to be followed by parentheses.

So from a parsing perspective, when each new keyword attribute is added, a binding decision has to be made about whether the keyword always takes arguments (and so parens are required) or never takes arguments (and so parens are immaterial).

The same comment applies to the other parsing changes as well.

sdesmalen-arm replied:
Addressed by https://github.com/llvm/llvm-project/pull/76971/commits/ecda4cd68b12b7938dfc2d047bc2aab08260cb09

rsandifo-arm commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -3175,11 +3175,16 @@ static void checkArmStreamingBuiltin(Sema &S, CallExpr *TheCall,
+  if (const auto *T = FD->getType()->getAs<FunctionProtoType>()) {
+    FunctionType::ArmStateValue State =
+        FunctionType::getArmZAState(T->getAArch64SMEAttributes());
+    if (State == FunctionType::ARM_In || State == FunctionType::ARM_Out ||
```
Shouldn't this include `ARM_Preserves` too?  `__arm_preserves("za")` functions are shared-ZA functions.

sdesmalen-arm replied:
Addressed by https://github.com/llvm/llvm-project/pull/76971/commits/e49a2dd0ef817eed3b6246a553efd7423b861130

rsandifo-arm commented on clang/lib/CodeGen/CGCall.cpp:
```diff
@@ -1767,14 +1767,22 @@ static void AddAttributesFromFunctionProtoType(ASTContext &Ctx,
+  if (FunctionType::getArmZAState(SMEBits) == FunctionType::ARM_Out ||
+      FunctionType::getArmZAState(SMEBits) == FunctionType::ARM_InOut)
+    FuncAttrs.addAttribute("aarch64_pstate_za_shared");
+  if (FunctionType::getArmZAState(SMEBits) == FunctionType::ARM_In) {
```
This is also the correct handling for `__arm_preserves("za")`.

sdesmalen-arm replied:
Addressed by https://github.com/llvm/llvm-project/pull/76971/commits/7e7bb75fdf49bc1c0917bdc5db9ac4ab0b5b67f4

rsandifo-arm reviewed (commented):
Thanks for doing this, looks good!  I've got some initial comments below, but I think they probably repeat for other parts of the patch.  The two main ones are:

* We should consume arguments to attribute keywords iff the attribute is defined to take arguments.
* Even `__arm_preserves("za")` functions are shared-ZA functions.

sdesmalen-arm committed:
Only parse arguments if those are required or optional [to squash]

sdesmalen-arm committed:
__arm_preserves("za") also shares ZA [to squash]

This also removes the functionality previously added to allow
dropping the '__arm_preserves_za' attribute in conversions,
since __arm_preserves("za") is no longer just a hint, it
is effecitvely instructing the compiler to share ZA.

sdesmalen-arm committed:
Rename 'PStateZA' -> 'ZA', and 'sme_state' to 'arm_state' [to squash]

sdesmalen-arm committed:
Include __arm_preserves("za") when testing if a function has ZA state [to quash]

sdesmalen-arm committed:
Rephrase undefined state 'S' in AttrDocs

sdesmalen-arm committed:
Appease the code formatting workflow [to squash]

rsandifo-arm reviewed (commented):
Thanks for the update, looks great to me.  The only substantive comment I have is that I think a given keyword attribute must always take `(` and `)` or never take them.  It shouldn't be optional, because that risks creating parsing ambiguities, or at least non-obvious behaviour.

sdesmalen-arm committed:
Make parsing of parentheses mandatory when an attribute keyword takes arguments [to squash]

sdesmalen-arm committed:
Use lambda to check for incompatible attributes [to squash]

sdesmalen-arm committed:
Change c2x-attribute-keywords.c to test both forms

sdesmalen-arm committed:
Fix up doc and tests [to squash]s

sdesmalen-arm commented on clang/test/Parser/c2x-attribute-keywords.c:
```diff
@@ -1,60 +1,60 @@
 
-enum __arm_streaming E { // expected-error {{'__arm_streaming' only applies to non-K&R-style functions}}
-  One __arm_streaming, // expected-error {{'__arm_streaming' only applies to non-K&R-style functions}}
+enum __arm_inout("za") E { // expected-error {{'__arm_inout' only applies to non-K&R-style functions}}
```
I've added explicit tests for unbalanced parentheses, but also changed this test to test both attributes (with and without arguments), using a macro that is sed-replaced with a RUN line.

sdesmalen-arm commented on clang/test/CodeGen/aarch64-sme-intrinsics/acle_sme_zero.c:
```diff
@@ -66,7 +66,7 @@ void test_svzero_mask_za_2(void) __arm_shared_za {
 // CHECK-CXX-NEXT:    ret void
 //
-void test_svzero_za(void) __arm_shared_za {
+void test_svzero_za(void) __arm_inout("za") {
```
Done.

sdesmalen-arm commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -8806,26 +8806,90 @@ static bool MustDelayAttributeArguments(const ParsedAttr &AL) {
+                                           const FunctionProtoType *FPT,
+                                           FunctionType::ArmStateValue State,
+                                           StringRef StateName) {
+  std::string ArmNewStr =
```
Done.

sdesmalen-arm commented on clang/include/clang/Basic/AttributeCommonInfo.h:
```diff
@@ -255,6 +255,22 @@ class AttributeCommonInfo {
   }
 };
+
+enum class KeywordAttributeParseArgumentsKind { None, Optional, Required };
```
Done.

sdesmalen-arm commented on clang/include/clang/Basic/AttrDocs.td:
```diff
@@ -6898,13 +6941,18 @@ at the end of the function.
+that the function will create a new scope for state S.
+
+The attribute takes string arguments to instruct the compiler which state
+is shared.  The supported states for S are:
```
Done.

sdesmalen-arm committed:
Appease the code formatting workflow again [to squash]

sdesmalen-arm commented on clang/test/CodeGen/aarch64-sme2-intrinsics/acle_sme2_write.c:
```diff
@@ -30,7 +30,7 @@
 // CPP-CHECK-NEXT:    ret void
 //
-void test_svwrite_ver_za8_u8_vg2(uint32_t base, svuint8x2_t val) __arm_streaming __arm_shared_za {
+void test_svwrite_ver_za8_u8_vg2(uint32_t base, svuint8x2_t val) __arm_streaming __arm_out("za") {
```
Done.

rsandifo-arm reviewed (approved):
LGTM.  There's one suggestion for a new test below, but please ignore if you think it's pointless.

sdesmalen-arm committed:
Add extra parser test

sdesmalen-arm committed:
Update cxx0x-keyword-attributes.cpp as well to test both forms [to squash]

Similar fashion as change to c2x-attribute-keywords.c.

sdesmalen-arm committed:
Fix up failing test that I missed previously [to squash]

sdesmalen-arm committed:
Fix up sed lines to appease Windows builds

Added quotes around the expression and replaced '/' by '@'. Not sure if
the latter is necessary, but I didn't see precedence for '/' in
existing tests.
